pkg/verifier/cosign/cosign.go

```diff
@@ -132,8 +186,161 @@ func (v *cosignVerifier) CanVerify(_ context.Context, referenceDescriptor ocispe
+}
+
+func (v *cosignVerifier) verifyInternal(ctx context.Context, subjectReference common.Reference, referenceDescriptor ocispecs.ReferenceDescriptor, referrerStore referrerstore.ReferrerStore) (verifier.VerifierResult, error) {
+	// get the map of keys and relevant cosign options for that reference
+	keysMap, cosignOpts, err := getKeysMaps(ctx, v.trustPolicies, subjectReference.Original, v.namespace)
+	if err != nil {
+		return errorToVerifyResult(v.name, v.verifierType, err), nil
+	}
+
+	// get the reference manifest (cosign oci image)
+	referenceManifest, err := referrerStore.GetReferenceManifest(ctx, subjectReference, referenceDescriptor)
+	if err != nil {
+		return errorToVerifyResult(v.name, v.verifierType, fmt.Errorf("failed to get reference manifest: %w", err)), nil
+	}
+
+	// manifest must be an OCI Image
+	if referenceManifest.MediaType != imgspec.MediaTypeImageManifest {
+		return errorToVerifyResult(v.name, v.verifierType, fmt.Errorf("reference manifest is not an image")), nil
+	}
+
+	// get the subject image descriptor
+	subjectDesc, err := referrerStore.GetSubjectDescriptor(ctx, subjectReference)
+	if err != nil {
+		return errorToVerifyResult(v.name, v.verifierType, fmt.Errorf("failed to create subject hash: %w", err)), nil
+	}
+
+	// create the hash of the subject image descriptor (used as the hashed payload)
+	subjectDescHash := v1.Hash{
+		Algorithm: subjectDesc.Digest.Algorithm().String(),
+		Hex:       subjectDesc.Digest.Hex(),
+	}
+
+	sigExtensions := make([]cosignExtensionList, 0)
+	hasValidSignature := false
+	// check each signature found
+	for _, blob := range referenceManifest.Blobs {
+		extensionListEntry := cosignExtensionList{
+			Signature:     blob.Annotations[static.SignatureAnnotationKey],
+			Verifications: make([]cosignExtension, 0),
+		}
+		// fetch the blob content of the signature from the referrer store
+		blobBytes, err := referrerStore.GetBlobContent(ctx, subjectReference, blob.Digest)
+		if err != nil {
+			return errorToVerifyResult(v.name, v.verifierType, fmt.Errorf("failed to get blob content: %w", err)), nil
+		}
+		// convert the blob to a static signature
+		staticOpts, err := staticLayerOpts(blob)
+		if err != nil {
+			return errorToVerifyResult(v.name, v.verifierType, fmt.Errorf("failed to parse static signature opts: %w", err)), nil
+		}
+		sig, err := static.NewSignature(blobBytes, blob.Annotations[static.SignatureAnnotationKey], staticOpts...)
+		if err != nil {
+			return errorToVerifyResult(v.name, v.verifierType, fmt.Errorf("failed to generate static signature: %w", err)), nil
+		}
+		// check each key in the map of keys returned by the trust policy
+		for mapKey, pubKey := range keysMap {
+			hashType := crypto.SHA256
+			// default hash type is SHA256 but for AKV scenarios, the hash type is determined by the key size
+			// TODO: investigate if it's possible to extract hash type from sig directly. This is a workaround for now
+			if pubKey.ProviderType == azurekeyvault.ProviderName {
```

Comment:
nit: probably we can extract this block into a single function.

Reply:
extracted to a helper function